backends/ebpf/runtime/ebpf_runtime.c

```diff
@@ -14,89 +14,270 @@ See the License for the specific language governing permissions and
+ and exposes an interface to feed packets to the emulated eBPF filter. */
 
-
-
-#include <string.h>
-#include <libgen.h>
+#include <unistd.h>     // getopt
+#include <ctype.h>      // isprint
 #define PCAP_DONT_INCLUDE_PCAP_BPF_H
 #include <pcap/pcap.h>
 #include "test.h"
+#ifdef CONTROL_PLANE
 #include "control.h"
-#include <unistd.h>
-#include <ctype.h>
+#endif
+
 #define FILE_NAME_MAX 256
+#define PCAPIN  "_in.pcap"
+#define PCAPOUT "_out.pcap"
+#define DELIM   "*"
 
 static int debug = 0;
 
 void usage(char *name) {
-    fprintf(stderr, "This program parses a pcap file, "
-            "feeds the individual packets into a filter function, "
+    fprintf(stderr, "This program expects a pcap file pattern, "
+            "extracts all the packets out of the matched files,"
+            "then feeds the individual packets into a filter function, "
             "and returns the output.\n");
-    fprintf(stderr, "Usage: %s [-d] -f file.pcap\n", name);
+    fprintf(stderr, "Usage: %s [-d] -f file.pcap [-n num_files]\n", name);
     fprintf(stderr, "Options:\n");
     fprintf(stderr, "\t-d: Turn on debug messages\n");
     fprintf(stderr, "\t-f: The input pcap file\n");
+    fprintf(stderr, "\t-n: Specifies the number of input files\n");
     exit(EXIT_FAILURE);
 }
 
-static int process_packets(pcap_t *in_handle, pcap_dumper_t *out_handle) {
-    struct pcap_pkthdr *pcap_hdr;
-    const unsigned char *packet;
-    int ret;
-    while ((ret = pcap_next_ex(in_handle, &pcap_hdr, &packet)) == 1) {
+/* Struct containing packet content and its metadata.
+ For now this is only the pcap header and the input interface.
+ */
+typedef struct {
+    char *data;
+    struct pcap_pkthdr pcap_hdr;
+    uint32_t ifindex;
+} pcap_pkt;
+
+/* Carries a list of packets and the total length as information.
+ Used as input stream for the packet parsing interface.
+ */
+typedef struct {
+    pcap_pkt *list;
+    uint64_t len;
+} pcap_pkt_list;
+
+/* Rank packets based on the timestamp of the pcap header */
+int compare(const void *s1, const void *s2) {
+  pcap_pkt *p1 = (pcap_pkt *)s1;
+  pcap_pkt *p2 = (pcap_pkt *)s2;
+  return p1->pcap_hdr.ts.tv_usec > p2->pcap_hdr.ts.tv_usec;
+}
+
+static pcap_pkt_list *sort_packets(pcap_pkt_list *pkt_lists, int num_lists) {
+    /* Get the total length of all individual lists combined first */
+    uint64_t total_len = 0;
+    for (int i = 0; i < num_lists; i++)
+        total_len+= pkt_lists[i].len;
+    /* Allocate the "master" list */
+    pcap_pkt_list *sorted_list = malloc(sizeof(pcap_pkt_list));
+    pcap_pkt *full_pkts = malloc(sizeof(pcap_pkt) * total_len);
+
+    /* Fill the master list by copying over the individual packets */
+    uint64_t offset = 0;
+    for (int i = 0; i < num_lists; i++) {
+        memcpy(&full_pkts[offset], pkt_lists[i].list, sizeof(pcap_pkt) *pkt_lists[i].len);
+        offset += pkt_lists[i].len;
+        /* We do not need the temporary allocated list anymore */
+        free(pkt_lists[i].list);
```

Comment:
In general you should delete something in the same place that allocated it, otherwise it is very confusing.
If a function deletes one of its arguments it should be clearly documented. But try to avoid doing this.